### ISSUE TITLE
Allow jax.pure_callback to infer result_shape_dtypes

### ISSUE THREAD

**Issue by carlosgmartin** [google/jax#16613]
I was wondering if it would be possible to allow [`jax.pure_callback`](https://jax.readthedocs.io/en/latest/_autosummary/jax.pure_callback.html) to infer `result_shape_dtypes` by calling the specified function once at trace time and using the result's treedef. This saves users the trouble of having to manually specify `result_shape_dtypes`. Possible implementation:

```python3
import jax
import numpy as np

def to_numpy(tree):
    return jax.tree_map(lambda x: np.empty(x.shape, x.dtype), tree)

def pure_callback_auto(f, *args, **kwargs):
    tree = f(*to_numpy(args), **to_numpy(kwargs))
    return jax.pure_callback(f, tree, *args, **kwargs)
```

**Comment by carlosgmartin** [comment]
@jakevdp Would you like me to create a PR for this?

**Comment by jakevdp** [comment]
I'm interested to hear what @sharadmv thinks of the idea

**Comment by carlosgmartin** [comment]
@jakevdp Now that `pure_callback` and `io_callback` are [switching from NumPy arrays to JAX arrays](https://github.com/google/jax/commit/9616900cc94ae9e795dd01b473fefd8315d99490), this seems even easier to do: Just call `jax.eval_shape(callback, *args, **kwargs)`. Is my understanding correct?

**Comment by jakevdp** [comment]
@carlosgmartin I think that would work in the simplest cases where the callback functions are implemented in terms of JAX operations. But I think this will be a rare case: why do a callback if your operations are entirely expressible in terms of JAX operations? More frequently the callback will likely call out to some other non-JAX API, for which `eval_shape` will fail.

**Comment by carlosgmartin** [comment]
Good point. I wonder if there's a way to maintain the zero-flop property (for JAX operations) while computing with non-JAX operations. There's still the possibility that `result_shape_dtypes` depends on the _value_ of the JAX input, but that seems like a rarer situation.

**Comment by carlosgmartin** [comment]
Some of the cases I had in mind were things like
```python3
import jax
from timeit import default_timer
jax.eval_shape(default_timer)
```
```
ShapeDtypeStruct(shape=(), dtype=float32)
```

**Comment by carlosgmartin** [comment]
I've deleted my previous comment, since I was confused.

I think [`jax.eval_shape`](https://jax.readthedocs.io/en/latest/_autosummary/jax.eval_shape.html) can work even if the function is not entirely expressible in terms of JAX operations. For example, consider [`timeit.default_timer`](https://docs.python.org/3/library/timeit.html#timeit.default_timer):
```python3
import timeit
import jax

print(jax.eval_shape(timeit.default_timer))
# ShapeDtypeStruct(shape=(), dtype=float32)
```

This allows one to do things like the following:
```python3
import timeit
import jax

def io_callback_auto(callback, *args, **kwargs):
    shape = jax.eval_shape(callback, *args, **kwargs)
    return jax.experimental.io_callback(callback, shape, *args, **kwargs)

def f(h, _):
    return h, io_callback_auto(timeit.default_timer)

_, xs = jax.lax.scan(f, None, None, length=10**6)

print(xs)
# [881338.3 881338.3 881338.3 ... 881340.  881340.  881340. ]
```

Could you give a concrete example where this wouldn't work, if you had one in mind?

**Comment by jakevdp** [comment]
I don’t disagree that some useful callbacks are compatible with `eval_shape`, I’m just pointing out that it’s likely the majority will not be. It’s not clear to me that it’s helpful to have a default that will fail in the majority of cases.
